### ISSUE TITLE
Settings table - condense the amount of code

### ISSUE THREAD

**Issue by dbashby** [thedigicraft/Atom.CMS#181]
I currently have all of the settings table outputting the data, site-title, debug and admin title.  It is all working through the use of classes however for the small amount of data I am retrieving I have a lot of code and I am trying to condense it into 1 prepared statement.

In settings.php I currently have 

`require 'classes.php';
	#Calls the settings class
	//debug panel 
	$id = "debug-status";
	$stmt = $pdo->prepare("SELECT * FROM settings WHERE id = :id");
	$stmt->bindParam(":id",$id,PDO::PARAM_STR);
	if($stmt->execute()){
		
		$rows = $stmt->fetchObject("Debug");
		$debug = $rows->get('value');
	}`

`//Site title
	$id = "site-title";
	$stmt = $pdo->prepare("SELECT * FROM settings WHERE id = :id limit 1");
	$stmt->bindParam(":id",$id,PDO::PARAM_STR);
	if($stmt->execute()){
		
		$rows = $stmt->fetchObject("Site_title");
		$siteTitle = $rows->get('value');
	}`

`//Admin title
	$id = "admin-title";
	$stmt = $pdo->prepare("SELECT * FROM settings WHERE id = :id limit 1");
	$stmt->bindParam(":id",$id,PDO::PARAM_STR);
	if($stmt->execute()){
		
		$rows = $stmt->fetchObject("Admin_title");
		$adminTitle = $rows->get('value');
	}`

I realise that as I am fetching an object I will only get 1 result but when I try to fetchAssoc I get nothing in return

In my classes.php I also have 3 sets of code each 1 just getting the 1 value

`#Debug Panel:
	class Debug
	{
		public function get($debug)
		{
			if(property_exists($this, $debug))
			{
				return $this->$debug;
			}
			return false;
		}
	}`

And bar the variables the other 2 pieces of code are the same returning the value for the debug panel, the site title and the admin site title.

There must be a way to condense this code into 1 far simpler prepared statement but I cannot find it.  I have had the array outputting the whole table with

`$stmt = $pdo->prepare("SELECT * FROM settings");
	$stmt->bindParam(1,$id,PDO::PARAM_STR);
	$stmt->execute();
		echo "<pre>";
	while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
		
		echo 'Echo Output: ' . 'id: ' . $row['id'] . ' ' . 'label: ' . $row['label'] . ' ' . 'value: ' . $row['value'] .'<br>';
		//echo var_dump($row);
		
	}
	echo "</pre>";
`

But I cannot seem to work out why I cannot retrieve all the data and then echo each element out.

![settingstable](https://cloud.githubusercontent.com/assets/6895380/13957550/b401a61e-f044-11e5-8956-cb73b422b13b.png)

**Comment by creptor** [comment]
I haven't look up if this is a correct way to handle this, but it will make your site faster, and your code less populated.

If you don't understand the code I have basically created a PDO call, that makes the id a variable storing it's value in the process.

Code:
```
function create_options(){
	$conn=$pdo->query('SELECT id,value FROM options');
	if($conn->rowCount()>0){
		foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
			global $$data['id'];
			$$data['id']=$data['value'];
		}
	}
}
create_options();
```
> Note: fetch object is intended to work with clases (`class`) in PHP, not in the way you're treating them.

**Comment by dbashby** [comment]
Unfortunately once I put this function in place of my previous code I get nothing returned when I try to reload the page.  I have looked through your code and can only see the $pdo not defined in the query so am I right in saying that that is all that needs to be in the brackets?

`function create_options($pdo){
}`

As well as `create_options($pdo);`

Thanks

**Comment by creptor** [comment]
you're right, sorry my mistake. It was taken from somewhere where I didn't need to define the `$pdo`.
Making that change should make it work.

**Comment by dbashby** [comment]
Unfortunately not. Nothing is returned I will take another look at it in the morning.

**Comment by creptor** [comment]
the idea of the script is that it doesn't return anything (no `retun` by the way), it just sets up the options variables.
To test it, you must comment the real variables, or then no noticiable changes will be made.

**Comment by dbashby** [comment]
Sorry it's been a while. So I just place the function inside the debug panel under the relevant listing and all the information should be displayed?

> create_options($pdo);

So instead of

`<?php echo '<pre>'; ?>
        <?php echo 'Database Debug Value Set: ' . $debug . '<br>'; ?>
        <?php echo 'Site Title: ' . $siteTitle . '<br>'; ?>
        <?php echo 'Admin Title: ' . $adminTitle . '<br>'; ?>
    <?php echo '</pre>'; ?>`

I should just have 

`create_options($pdo);`

Is this correct?

**Comment by creptor** [comment]
no, the create_options function that I gave you only sets the variable and values, it **does nothing else**, you would still need them where you need them.

Example:
```
//in database => ('id' => 'title' , 'value' => 'My site')
create_options($pdo);
//returns $title='My site';
echo $tile
//displays the $title variable
```

**Comment by dbashby** [comment]
I am sorry but I must be missing something, when I try to use the function I get a blank page returned. My understanding from the above is that say on the index page.

<title>echo $pageTitle | $siteTitle</title> is no good now

I need

<title> ? </title>

As I say I have tried all sorts, even putting the function, create_options($pdo); between the title tags but I do not get anything.

**Comment by dbashby** [comment]
Just looking at the code again, do I need to declare $data in the function brackets as well?

`function create_options($pdo, $data){
		$conn=$pdo->query('SELECT id,value FROM settings');
		if($conn->rowCount()>0){
			foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
				global $$data['id'];
				$$data['id']=$data['value'];
			}
		}
	}`

i.e. create_options($pdo, $data);

Although I have tried this and still only blank page returned.

**Comment by creptor** [comment]
no, $data is not necessary, please add a `error_reporting(-1);`

**Comment by dbashby** [comment]
The error returned

> Parse error: syntax error, unexpected '[', expecting ',' or ';' in >C:\xampp\htdocs\1mysite\php_inc\setup.php on line 46

Line 46 is

`global $$data['id'];`

of the function

`function create_options($pdo, $data){
		$conn=$pdo->query('SELECT id,value FROM settings');
		if($conn->rowCount()>0){
			foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
				global $$data['id'];
				$$data['id']=$data['value'];
			}
		}
	}`

**Comment by creptor** [comment]
please use pastebin to upload your setup.php

**Comment by dbashby** [comment]
I have put in new file due to volume of code

http://pastebin.com/6jZFaArU

**Comment by creptor** [comment]
no, 404 error is synonymous of `not found`, and `file_exists()` check if the file does exists, which is not the same. also they use different check functions, a 404 error is for html headers, and not system files (it can still work dough).

Also be cautious, because your PHP server may cache the result of this function, resulting in misleading results.


This also has some noticeable problems like this one:
> Warning
This function returns FALSE for files inaccessible due to safe mode restrictions. However these files still can be included if they are located in safe_mode_include_dir.

Quoted from php.net

For checking URIs response is better to get the response headers, use a code like this one (there're many ways to do this).
```
$file = 'http://www.domain.com/somefile.jpg';
$file_headers = @get_headers($file);
if($file_headers[0] == 'HTTP/1.1 404 Not Found') {
    $exists = false;
}
else {
    $exists = true;
}
```

**Comment by creptor** [comment]
I have found out why my code didn't work, it was the variables....

Example:
```
//variable name needed `$siteTitle` (for the title)

**->execution of my function

//variable gotten: `$site-title`
```

Note that `$site-title` is different than `$siteTitle`, which is why you're getting some blank stuff. To fix this, you have to change (in the database) the `id` column, to match the needed variable name.

That's what basically my code does, makes the id string a variable, and it inserts the value to it :3

**Comment by creptor** [comment]
Also my code has no error... I have it working, and that error makes no sense.

Try checking your included files, this page works like a charm http://phpcodechecker.com/

**Comment by dbashby** [comment]
![settingsdb](https://cloud.githubusercontent.com/assets/6895380/14194809/3eed7408-f7aa-11e5-9d68-cf66b9bae09d.png)

>To fix this, you have to change (in the database) the id column, to match the needed variable >name.

As you see the column name in the database 'id' for the title is site-title, so it is the variables in the pages I need to correct to fit yes? 

I do appreciate the time your spending on this, so thanks

**Comment by dbashby** [comment]
I have changed the appropriate variables to match with the DB 

![settingsdb](https://cloud.githubusercontent.com/assets/6895380/14195665/ea5ce9dc-f7b3-11e5-8724-0b50bc3fad83.png)

but still getting the

>Parse error: syntax error, unexpected '[', expecting ',' or ';' in >C:\xampp\htdocs\1mysite\php_inc\setup.php on line 46

**Comment by dbashby** [comment]
The only 2 pages being called in are

include 'connection/connect.php';
	require 'functions.php';

We know the connection is working along with functions.php

**Comment by dbashby** [comment]
I have found 1 error message, I had to force the error though.  I just added random letters to the url so

>http://localhost/1mysite/index.php?page=home

became

>http://localhost/1mysite/index.php?page=homefggd

The error I got was 

>Warning: Invalid argument supplied for foreach() in C:\xampp\htdocs\1mysite\php_inc\setup.php on >line 68

That is the foreach loop in the function

>foreach($page as $row){

`$page = data_page($pdo, $pageid); 
	foreach($page as $row){
    	$data=$row;
		//var_dump($data);
	}
		return $data;`

Could this be why the function is not working even though they are not connected as 1 is dealing with the settings table and the other is working with the posts table?

I also loose the page information in the tab, obviously thats to be expected as the page
>http://localhost/1mysite/index.php?page=homefggd
does not exist

At the same time I also got
>Undefined variable: data in C:\xampp\htdocs\1mysite\php_inc\setup.php on line 72

But again I cannot see how that would stop the function below from working?

`function create_options(){
    $conn=$pdo->query('SELECT id,value FROM settings');
    if($conn->rowCount()>0){
        foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
            global $$data['id'];
            $$data['id']=$data['value'];
        }
    }
}
create_options();`

**Comment by creptor** [comment]
Alright, I'll check if it's necessary to add something else to my function to work properly, and which php version is needed to run it.

**Comment by dbashby** [comment]
Just so you know, I am running 7.0.3

**Comment by creptor** [comment]
I'm sorry I think I have lost it... please upload again an image of how your site looks, and the functions file (please separate them, that way the line numbers match).

**Comment by dbashby** [comment]
Ok, no problem.  Here are all the pages in seperate files.
http://pastebin.com/29WNw1jR - Index
http://pastebin.com/gE9UvqAe - Functions
http://pastebin.com/SkkrD9Du - setup
http://pastebin.com/PL6fiMjc - Config - not much in this

I think that is everything, let me know if you need anything else.  Thanks

![poststable](https://cloud.githubusercontent.com/assets/6895380/14221103/67d21e02-f85c-11e5-9a78-8e93c9a5cd0e.png)

**Comment by creptor** [comment]
I didn't explain myself properly. With the image I was referring to how the site look, that way I can find error easier. Or at least know the file where they are.

**Comment by dbashby** [comment]
![mobileview](https://cloud.githubusercontent.com/assets/6895380/14225739/65b46bde-f8c5-11e5-8eab-2e2c62f50323.png)

![desktopview](https://cloud.githubusercontent.com/assets/6895380/14225740/697c81e8-f8c5-11e5-9d47-d1d62678df4f.png)

**Comment by creptor** [comment]
The images show no site errors, do you still ave some of them?, please if so copy the hole error output.

**Comment by dbashby** [comment]
The errors come up due to the function that gets all the settings from the settings table.  The error I get is 
>Parse error: syntax error, unexpected '[', expecting ',' or ';' in >C:\xampp\htdocs\1mysite\php_inc\setup.php on line 46

This relates to the 

`#Settings table function
		function create_options($pdo, $data){
		$conn=$pdo->query('SELECT id,value FROM settings');
		if($conn->rowCount()>0){
			foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
				global $$data['id'];
				$$data['id']=$data['value'];
				}
			}
		}
		create_options($pdo);`

**Comment by creptor** [comment]
try replacing my function with this:

```
function create_options($pdo){
    $conn=$pdo->query('SELECT id,value FROM settings');
    if($conn->rowCount()>0){
        foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
            $key=$data['id']
            global $$key;
            $$key=$data['value'];
        }
    }
}
create_options($pdo);
```

Also, try to comment out the `foreach($page as $row){ $data=$row; } return $data;`, to see if something gets broken. If is that the case please tell me what is it, if not just leave it that way :+1:

**Comment by dbashby** [comment]
I replaced the old function with the new, I got an error

>Parse error: syntax error, unexpected 'global' (T_GLOBAL) in >C:\xampp\htdocs\1mysite\php_inc\functions.php on line 67

![error1](https://cloud.githubusercontent.com/assets/6895380/14233377/99323528-f9be-11e5-9277-7986ba3eb931.png)

**Comment by creptor** [comment]
.-. , maybe is the Apache configuration, here this code should work, but is more extensive than the last, and could result in unexpected behavior.

```
function create_options($pdo){
    $conn=$pdo->query('SELECT id,value FROM settings');
    if($conn->rowCount()>0){
        $return=array();
        foreach($conn->fetchAll(PDO::FETCH_ASSOC) as $data){
            $return[]=array($data['id'],$data['value']);
        }
        return $return;
    }
}
//returns array(x){'0'=>array(2)('the_id','the_value'), ..... }
$options=create_options($pdo);
foreach($options as $data=>$value){
    $$value[0]=$value[1];
}
```

**Comment by dbashby** [comment]
Unfortunately not, I get the following and as you see I lose the siteTitle, I realise that due to the error I would but just pointing it out.

![arraytostring](https://cloud.githubusercontent.com/assets/6895380/14250292/a7f21b16-fa75-11e5-894b-13424326e1ba.png)

![arraytostring2](https://cloud.githubusercontent.com/assets/6895380/14250300/adfed26a-fa75-11e5-942a-c591062a0bca.png)

**Comment by dbashby** [comment]
Has anyone come up with a fix for this?  Thanks

**Comment by creptor** [comment]
maybe try:
```
$$value[$data][0]=$value[$data][1];
```
